File: rust/replicant/src/replicant/multipaxos/msg.rs
use serde::{Deserialize, Serialize};
use crate::replicant::multipaxos::tcp::Message;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ResponseType {
    Ok = 0,
    Reject = 1,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
#[repr(i32)]
pub enum CommandType {
    Get = 0,
    Put = 1,
    Del = 2,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum InstanceState {
    Inprogress = 0,
    Committed = 1,
    Executed = 2,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Command {
    pub r#type: i32,
    pub key: String,
    pub value: String,
}

#[derive(Clone, PartialEq, Serialize, Deserialize)]
pub struct Instance {
    pub ballot: i64,
    pub index: i64,
    pub client_id: i64,
    pub state: i32,
    pub command: Option<Command>,
}

#[derive(Serialize, Deserialize)]
pub struct PrepareRequest {
    pub ballot: i64,
    pub sender: i64,
}

#[derive(Serialize, Deserialize)]
pub struct PrepareResponse {
    pub r#type: i32,
    pub ballot: i64,
    pub instances: Vec<Instance>,
}

#[derive(Serialize, Deserialize)]
pub struct AcceptRequest {
    pub instance: Option<Instance>,
    pub sender: i64,
}

#[derive(Serialize, Deserialize)]
pub struct AcceptResponse {
    pub r#type: i32,
    pub ballot: i64,
}

#[derive(Serialize, Deserialize)]
pub struct CommitRequest {
    pub ballot: i64,
    pub last_executed: i64,
    pub global_last_executed: i64,
    pub sender: i64,
}

#[derive(Serialize, Deserialize)]
pub struct CommitResponse {
    pub r#type: i32,
    pub ballot: i64,
    pub last_executed: i64,
}

#[derive(Copy, Clone)]
#[repr(u8)]
pub enum MessageType {
    PrepareRequest = 1,
    PrepareResponse,
    AcceptRequest,
    AcceptResponse,
    CommitRequest,
    CommitResponse,
}

File: go/multipaxos/network/msg.go
package network

type ResponseType int32

const (
	Ok     ResponseType = 0
	Reject ResponseType = 1
)

type CommandType int32

const (
	Get CommandType = 0
	Put CommandType = 1
	Del CommandType = 2
)

type InstanceState int32

const (
	Inprogress InstanceState = 0
	Committed  InstanceState = 1
	Executed   InstanceState = 2
)

type MessageType uint8

const (
	PREPAREREQUEST MessageType = iota + 1
	PREPARERESPONSE
	ACCEPTREQUEST
	ACCEPTRESPONSE
	COMMITREQUEST
	COMMITRESPONSE
)

type Command struct {
	Type  CommandType
	Key   string
	Value string
}

type Instance struct {
	Ballot   int64
	Index    int64
	ClientId int64
	State    InstanceState
	Command  *Command
}

type PrepareRequest struct {
	Ballot int64
	Sender int64
}

type PrepareResponse struct {
	Type   ResponseType
	Ballot int64
	Logs   []*Instance
}

type AcceptRequest struct {
	Instance *Instance
	Sender   int64
}

type AcceptResponse struct {
	Type   ResponseType
	Ballot int64
}

type CommitRequest struct {
	Ballot             int64
	LastExecuted       int64
	GlobalLastExecuted int64
	Sender             int64
}

type CommitResponse struct {
	Type         ResponseType
	Ballot       int64
	LastExecuted int64
}
